lib/pbench/cli/server/shell.py

```diff
@@ -25,19 +29,71 @@ def app():
     return create_app(server_config)
 
 
+def add_to_path(path_el: str, path: str) -> str:
+    """Convenience function to hide f-string syntax used to construct a new
+    PATH environment variable.
+
+    Arguments:
+
+        path_el - a new path element string to be pre-pended to `path`
+        path - the exiting PATH environment variable value
+
+    Returns:
+
+        A string representing the newly constructed PATH value
+    """
+    return f"{path_el}:{path}"
+
+
 def find_the_unicorn(logger: Logger):
-    if site.ENABLE_USER_SITE:
-        local = Path(site.getuserbase()) / "bin"
-        if (local / "gunicorn").exists():
-            # Use a `pip install --user` version of gunicorn
-            os.environ["PATH"] = str(local) + ":" + os.environ["PATH"]
-            logger.info(
-                "Found a local unicorn: augmenting server PATH to {}",
-                os.environ["PATH"],
-            )
+    """Add the location of the `pip install --user` version of gunicorn to the
+    PATH if it exists.
+    """
+    local = Path(site.getuserbase()) / "bin"
+    if (local / "gunicorn").exists():
+        # Use a `pip install --user` version of gunicorn
+        os.environ["PATH"] = add_to_path(local, os.environ["PATH"])
+        logger.info(
+            "Found a local unicorn: augmenting server PATH to {}",
+            os.environ["PATH"],
+        )
+
+
+def wait_for_postgresql(db_uri: str, timeout: int):
+    """Wait for PostgreSQL database server to become available.  While we
+    encounter "connection refused", sleep one second, and then try again.
+
+    The timeout argument to `create_connection()` does not play into the retry
+    logic, see:
+
+      https://docs.python.org/3.9/library/socket.html#socket.create_connection
+
+    Arguments:
+
+        timeout: integer number of seconds to wait before giving up attempts to
+                 connect to the database
+
+    Raises a ConnectionRefusedError on failure.
+    """
+    url = urlparse(db_uri)
+    port = url.port if url.port is not None else 5432
+    while timeout >= 0:
+        try:
+            with socket.create_connection((url.hostname, port), timeout=1):
+                break
+        except ConnectionRefusedError:
+            time.sleep(1)
+            timeout -= 1
 
 
 def main():
+    """The Pbench Server is always executed in a containerized environment where
+    the pbench server configuration is in a fixed location.
```

Comment:
I'm not sure that containerization has anything to do with anything, here.  And, I'm not convinced that we want to claim to depend on running in/as a container.

The point to be made here is that the configuration file is in a known location.  It's too bad that we have to define it here, rather than pulling it in from a single point of truth (although, I guess it is possible that this _**is**_ the single point of truth for this...).

In any case, I would omit this whole sentence from the docstring, and perhaps add a description of what this function actually does (e.g., reads the server configuration, waits for the database service to become available, create the database, and so on)...or just a generic comment like, "Starts the Pbench Server".